sams.go

```diff
@@ -0,0 +1,59 @@
+package sams
+
+import "github.com/sourcegraph/sourcegraph/lib/errors"
+
+// envGetter is a helper interface for getting environment variables based on
+// the MSP runtime environment, matching the type
+// github.com/sourcegraph/sourcegraph/lib/managedservicesplatform/runtime/contract.Env.
+type envGetter interface {
+	// Get returns the value with the given name. If no value was supplied in the
+	// environment, the given default is used in its place. If no value is available,
+	// an error is added to the validation errors list.
+	Get(name, defaultValue, description string) string
+	// GetOptional returns the value with the given name, or nil if no value is
+	// available.
+	GetOptional(name, description string) *string
+}
+
+// ConnConfig is the basic configuration for connecting to a Sourcegraph Accounts
+// instance. Callers SHOULD use sams.NewConnConfigFromEnv(...) to construct this
+// where possible to ensure connection configuration is unified across SAMS clients
+// for ease of operation by Core Services team.
+type ConnConfig struct {
+	// ExternalURL is the configured default external ExternalURL of the relevant Sourcegraph
+	// Accounts instance.
+	ExternalURL string
+	// APIURL is the URL to use for Sourcegraph Accounts API interactions. This
+	// can be set to some internal URLs for private networking. If this is nil,
+	// the client will fall back to ExternalURL instead.
+	APIURL *string
+}
+
+const DefaultExternalURL = "https://accounts.sourcegraph.com"
+
+// NewConnConfigFromEnv initializes configuration for connecting to Sourcegraph
+// Accounts using default standards for loading environment variables. This
+// allows the Core Services team to more easily configure access.
+func NewConnConfigFromEnv(env envGetter) ConnConfig {
+	return ConnConfig{
+		ExternalURL: env.Get("SAMS_URL", DefaultExternalURL, "External URL of the connected SAMS instance"),
+		APIURL:      env.GetOptional("SAMS_API_URL", "URL to use for connecting to the API of a SAMS instance instead of SAMS_URL"),
+	}
+}
+
+func (c ConnConfig) Validate() error {
```

Comment:
I don't know about the requirements around `APIURL` and private networking. But one thing that the new "accounts client" requires is that the `ExternalURL` does _not_ end in a trailing slash. e.g. "http://localhost:9992" NOT "http://localhost:9992/". This is only because we generate URLs by concatenating strings, and aren't canonicalizing/normalizing URL paths to account for potential problems like "...localhost:9992//api/...".

Reply:
This change uses the same constructor under-the-hood in `accounts/v1`, so the URL-cleaning behaviour is retained. I think simple changes like this should be elegantly handled by constructors - forgetting to remove the trailing slash is an annoying reason to have your service fail to start in production 😆